### DISCUSSION TITLE
Filter Self Relationship

### DISCUSSION THREAD

**Discussion by Aetylus** [sqlalchemy/sqlalchemy#8633] [Usage Questions]
I have the following model with a self referenced relationship:

```py
class Node(Base):
  id = Column(Integer, Identity(), primary_key=True)
  parent_id = Column(Integer, ForeignKey('node.id')
  active = Column(Boolean)
  
  children = relationship('Node', backref('parent', remote_side=[id])
```

How would I query this such that I get all `Node` records that have `active == true` but only receive the `children` that also have `active == true`?

I've tried doing the following with the relationship:

```py
children = relationship(
  'Node',
  primaryjoin='and_(remote(Node.parent_id) == foreign(Node.id), remote(Node.active) == true())',
  backref=backref('parent', remote_side=[id])
)
```

However, this ends up with the resulting query not returning a list of `children` and instead only returning a single value for reasons I couldn't decipher.

I've also tried doing it on the query side:

```py
nodes = db.query(Node).filter(Node.active == true(), Node.children.any(Node.active == true()))
```

However, this only returns back `Node` records that *have* children that are active and fails to return back any `Node` records that do not have any children at all.

I've tried several other attempts at solving this but to no avail so far.

**Comment by zzzeek** [comment]
There's a few ways, but the most likely here would be to use the [with_loader_criteria](https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=with_loader_criteria#sqlalchemy.orm.with_loader_criteria) option that is meant for flags like "deleted", "active", etc:

    from sqlalchemy.orm import with_loader_criteria
    nodes = db.query(Node).options(with_loader_criteria(Node, Node.active == True, include_aliases=True))
